### PR TITLE
New client architecture

### PR DESCRIPTION
**Pull request by tdudzik**
Refactor of an architecture on client side. Now it works in this way:

Event producers: BattleshipsController, MessageProcessor
Event consumers: BattleshipsController, EventHandler

1. EventHandler handles an event from GUI and passes it to EventProcessor.
2. EventProcessor creates Message from handled event and pass it to MessageProcessor asynchronously using AsyncTask.
3. MessageProcessor sends Message using MessageSender and it's waiting for response.
4. When MessageProcessor gets response it posts an event to EventBus. Type of event depends on the response.

### CODE REVIEW

File: client/src/main/java/io/github/expansionteam/battleships/MainLauncher.java [hunk 1]
```diff
@@ -21,7 +20,7 @@ public void start(Stage primaryStage) throws Exception {
         Injector injector = Guice.createInjector(new BattleshipsModule());
         EventBus eventBus = injector.getInstance(EventBus.class);
 
-        eventBus.register(injector.getInstance(EventHandler.class));
+        eventBus.register(injector.getInstance(io.github.expansionteam.battleships.logic.event.EventHandler.class));
```
Comment (clinton56) [inline client/src/main/java/io/github/expansionteam/battleships/MainLauncher.java]: Please fix import.

File: client/src/main/java/io/github/expansionteam/battleships/gui/controllers/BattleshipsController.java [hunk 2]
```diff
@@ -36,8 +40,8 @@
     @FXML
     VBox playerBoardArea;
 
-    private OpponentBoard opponentBoard;
```
Comment (clinton56) [inline client/src/main/java/io/github/expansionteam/battleships/gui/controllers/BattleshipsController.java]: Why did you change private to package access>

**Comment by YogAzathoth** [comment]
Commits messages may have been a little more descriptive, but overall I like it

File: client/src/main/java/io/github/expansionteam/battleships/gui/controllers/BattleshipsController.java (hunk 2 above)
Comment (tdudzik) [inline client/src/main/java/io/github/expansionteam/battleships/gui/controllers/BattleshipsController.java]: We need package-private access in order to be able to mock it.

File: client/src/main/java/io/github/expansionteam/battleships/gui/models/Ship.java [hunk 3]
```diff
@@ -42,4 +43,18 @@ private static Ship create(ShipOrientation orientation, Position position, ShipS
...
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if ((o instanceof Ship) == false) {
+            return false;
+        }
+
+        Ship other = (Ship) o;
+        return Objects.equals(this.fieldsByPosition, other.fieldsByPosition);
```
Comment (clinton56) [inline client/src/main/java/io/github/expansionteam/battleships/gui/models/Ship.java]: Why do use `Objects.equals()`? I'm not sure if it will work correctly with collections. I would suggest to use `Objects.deepEquals()` or just `return fieldsByPosition.equals(other.fieldsByPosition)`, since both `Position` and `Field` are implementing `hashCode()` and `equals()`

File: client/src/main/java/io/github/expansionteam/battleships/logic/message/Message.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package io.github.expansionteam.battleships.logic.message;
+
+public class Message {
```
Comment (clinton56) [inline client/src/main/java/io/github/expansionteam/battleships/logic/message/Message.java]: I'm wondering whether Message is used outside of a package>?

File: client/src/main/java/io/github/expansionteam/battleships/logic/message/MessageSender.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Inject
+    public MessageSender(SocketChannel socketChannel, MessageFactory messageFactory) {
+        this.socketChannel = socketChannel;
+        this.messageFactory = messageFactory;
+    }
+
+    public Message sendMessageAndWaitForResponse(Message message) {
+        try {
+            DataInputStream inputStream = new DataInputStream(socketChannel.socket().getInputStream());
```
Comment (clinton56) [inline client/src/main/java/io/github/expansionteam/battleships/logic/message/MessageSender.java]: [Lines 27-28] If we don't use `try-with-resources` so our resources are not automatically closed and we have memory leak? If so parentheses are missing.

File: client/src/main/java/io/github/expansionteam/battleships/logic/message/MessageSender.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            DataOutputStream outputStream = new DataOutputStream(socketChannel.socket().getOutputStream());
+
+            outputStream.writeUTF(message.getBody());
+            outputStream.flush();
+
+            String json = inputStream.readUTF();
+            return messageFactory.createFromJson(json);
+        } catch (EOFException e) {
```
Comment (clinton56) [inline client/src/main/java/io/github/expansionteam/battleships/logic/message/MessageSender.java]: First thing: pipe could used.
Second: we shouldn't `printStackTrace()` to the console.

File: client/src/main/java/io/github/expansionteam/battleships/gui/models/Ship.java (hunk 3 above)
Comment (tdudzik) [inline client/src/main/java/io/github/expansionteam/battleships/gui/models/Ship.java]: `Objects.equals()` and `fieldsByPosition.equals(other.fieldsByPosition)` in this case are equivalent.

File: client/src/main/java/io/github/expansionteam/battleships/logic/message/Message.java (hunk 4 above)
Comment (tdudzik) [inline client/src/main/java/io/github/expansionteam/battleships/logic/message/Message.java]: We use it also in different packages (e. g. `event` package).

File: client/src/main/java/io/github/expansionteam/battleships/logic/message/MessageSender.java (hunk 5 above)
Comment (tdudzik) [inline client/src/main/java/io/github/expansionteam/battleships/logic/message/MessageSender.java]: `try-with-resources` in this case would close also our socket.

File: client/src/main/java/io/github/expansionteam/battleships/logic/message/MessageSender.java (hunk 6 above)
Comment (tdudzik) [inline client/src/main/java/io/github/expansionteam/battleships/logic/message/MessageSender.java]: We need better exceptions handling in this place. It will be added in the future.

File: client/src/main/java/io/github/expansionteam/battleships/gui/controllers/BattleshipsController.java (hunk 2 above)
Comment (LIttleAncientForestKami) [inline client/src/main/java/io/github/expansionteam/battleships/gui/controllers/BattleshipsController.java]: Are you certain?

AFAIK Mockito annotations use reflections under the hood. Also, you may introduce a test constructor.

File: client/src/main/java/io/github/expansionteam/battleships/gui/controllers/BattleshipsController.java [hunk 7]
```diff
@@ -48,6 +52,8 @@ public void initialize(URL location, ResourceBundle resources) {
         playerBoardArea.getChildren().add(playerBoard);
 
         boardArea.setVisible(false);
+
+        log.debug("Post StartGameEvent.");
```
Comment (LIttleAncientForestKami) [inline client/src/main/java/io/github/expansionteam/battleships/gui/controllers/BattleshipsController.java]: Debug -> Trace
